src/main/kotlin/org/jetbrains/kotlinx/jupyter/repl/impl/InternalEvaluatorImpl.kt

```diff
@@ -57,9 +67,18 @@ internal class InternalEvaluatorImpl(
 
     override val lastClassLoader get() = compiler.lastClassLoader
 
+    override val variablesHolder = mutableMapOf<String, VariableState>()
+
+    override val varsUsagePerCell = mutableMapOf<Int, MutableSet<String>>()
+
+    /**
+     * Tells in which cell a variable was declared
+     */
+    private val varsDeclarationInfo: MutableMap<String, Int> = mutableMapOf()
```

Comment:
`varsUsagePerCell` and `varsDeclarationInfo` is actually a bidirectional map, right? You may write a generic utility class that incapsulates the logic of working with them.
By the way, you've added 4 properties here for a single feature. It would be great if their number might be reduced (i.e. using incapsualtion to another class(es))

Reply:
It is indeed. Sounds like an interesting thing, would try

Reply:
Although it's not a bidirectional map in its original sense. We do want to have a supplementary map of declarations, so it's more like String -> Int with Int ->Set<String>, hence, types are not symmetric. It took me a while but in a meanwhile I've got the bidirectional generic map))